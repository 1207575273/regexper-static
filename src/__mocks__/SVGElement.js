import React from 'react';

import Base from 'components/SVG/Base';

class SVGElement extends Base {
  reflow() {
    return this.setBBox(this.props.bbox);
  }

  render() {
    return <text>Mock content</text>;
  }
}

export default SVGElement;
