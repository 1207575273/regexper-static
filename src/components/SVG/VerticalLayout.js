import React from 'react';
import PropTypes from 'prop-types';
import { List } from 'immutable';

import style from './style';

import reflowable from './reflowable';
import Path from './path';

const connectorMargin = 20;

@reflowable
class VerticalLayout extends React.PureComponent {
  static defaultProps = {
    withConnectors: false,
    spacing: 10
  }

  state = {
    childTransforms: List()
  }

  updateChildTransforms(childBoxes) {
    return this.state.childTransforms.withMutations(transforms => (
      childBoxes.forEach((box, i) => (
        transforms.set(i, `translate(${ box.offsetX } ${ box.offsetY })`)
      ))
    ));
  }

  makeCurve(box) {
    const thisBox = this.getBBox();
    const distance = Math.abs(box.offsetY + box.axisY - thisBox.axisY);

    if (distance >= 15) {
      const curve = (box.axisY + box.offsetY > thisBox.axisY) ? 10 : -10;

      return new Path()
        // Left
        .moveTo({ x: 10, y: box.axisY + box.offsetY - curve })
        .quadraticCurveTo({ cx: 0, cy: curve, x: 10, y: curve, relative: true })
        .lineTo({ x: box.offsetX + box.axisX1 })
        // Right
        .moveTo({ x: thisBox.width - 10, y: box.axisY + box.offsetY - curve })
        .quadraticCurveTo({ cx: 0, cy: curve, x: -10, y: curve, relative: true })
        .lineTo({ x: box.offsetX + box.axisX2 });
    } else {
      const anchor = box.offsetY + box.axisY - thisBox.axisY;

      return new Path()
        // Left
        .moveTo({ x: 0, y: thisBox.axisY })
        .cubicCurveTo({ cx1: 15, cy1: 0, cx2: 10, cy2: anchor, x: 20, y: anchor, relative: true })
        .lineTo({ x: box.offsetX + box.axisX1 })
        // Right
        .moveTo({ x: thisBox.width, y: thisBox.axisY })
        .cubicCurveTo({ cx1: -15, cy1: 0, cx2: -10, cy2: anchor, x: -20, y: anchor, relative: true })
        .lineTo({ x: box.offsetX + box.axisX2 });
    }
  }

  makeSide(box) {
    const thisBox = this.getBBox();
    const distance = Math.abs(box.offsetY + box.axisY - thisBox.axisY);

    if (distance >= 15) {
      const shift = (box.offsetY + box.axisY > thisBox.axisY) ? 10 : -10;
      const edge = box.offsetY + box.axisY - shift;

      return new Path()
        // Left
        .moveTo({ x: 0, y: thisBox.axisY })
        .quadraticCurveTo({ cx: 10, cy: 0, x: 10, y: shift, relative: true })
        .lineTo({ y: edge })
        // Right
        .moveTo({ x: thisBox.width, y: thisBox.axisY })
        .quadraticCurveTo({ cx: -10, cy: 0, x: -10, y: shift, relative: true })
        .lineTo({ y: edge });
    }
  }

  reflow() {
    const { spacing, withConnectors } = this.props;

    const childBoxes = this.children.map(child => child.getBBox());
    const horizontalCenter = childBoxes.reduce((center, box) => Math.max(center, box.width / 2), 0);
    const margin = withConnectors ? connectorMargin : 0;
    const width = childBoxes.reduce((width, box) => Math.max(width, box.width), 0) + 2 * margin;
    const height = childBoxes.reduce((height, box) => height + box.height, 0) + (childBoxes.length - 1) * spacing;
    this.setBBox({ width, height }, { axisY: true, axisX1: true, axisX2: true });

    let offset = 0;
    childBoxes.forEach(box => {
      box.offsetX = horizontalCenter - box.width / 2 + margin;
      box.offsetY = offset;
      offset += spacing + box.height;
    });

    this.setStateAsync({
      childTransforms: this.updateChildTransforms(childBoxes),
      connectorPaths: withConnectors ? [
        ...childBoxes.map(box => this.makeCurve(box)),
        this.makeSide(childBoxes[0]),
        this.makeSide(childBoxes[childBoxes.length - 1])
      ].join('') : ''
    });
  }

  childRef = i => child => this.children[i] = child

  render() {
    const { children } = this.props;
    const { childTransforms, connectorPaths } = this.state;

    this.children = [];

    return <React.Fragment>
      <path d={ connectorPaths } style={ style.connectors }></path>
      { React.Children.map(children, (child, i) => (
        <g transform={ childTransforms.get(i) }>
          { React.cloneElement(child, {
            ref: this.childRef(i)
          }) }
        </g>
      )) }
    </React.Fragment>;
  }
}

VerticalLayout.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
  spacing: PropTypes.number,
  withConnectors: PropTypes.bool
};

export default VerticalLayout;
