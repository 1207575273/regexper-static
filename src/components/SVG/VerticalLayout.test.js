import React from 'react';
import { mount } from 'enzyme';

import VerticalLayout from './VerticalLayout';

import SVGElement from '__mocks__/SVGElement';

describe('VerticalLayout', () => {
  test('rendering', async () => {
    const component = mount(
      <VerticalLayout>
        <SVGElement bbox={{ width: 100, height: 100 }}/>
        <SVGElement bbox={{ width: 100, height: 100 }}/>
        <SVGElement bbox={{ width: 100, height: 100 }}/>
      </VerticalLayout>
    );
    await component.instance().doReflow();
    component.update();
    expect(component).toMatchSnapshot();
  });

  test('rendering with connectors', async () => {
    const component = mount(
      <VerticalLayout withConnectors>
        <SVGElement bbox={{ width: 100, height: 100 }}/>
        <SVGElement bbox={{ width: 100, height: 100 }}/>
        <SVGElement bbox={{ width: 100, height: 100 }}/>
      </VerticalLayout>
    );
    await component.instance().doReflow();
    component.update();
    expect(component).toMatchSnapshot();
  });

  test('rendering with connectors (no sides)', async () => {
    const component = mount(
      <VerticalLayout withConnectors>
        <SVGElement bbox={{ width: 100, height: 10 }}/>
        <SVGElement bbox={{ width: 100, height: 10 }}/>
      </VerticalLayout>
    );
    await component.instance().doReflow();
    component.update();
    expect(component).toMatchSnapshot();
  });
});
