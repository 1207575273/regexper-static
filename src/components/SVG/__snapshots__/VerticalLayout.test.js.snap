// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerticalLayout rendering 1`] = `
<VerticalLayout
  spacing={10}
  withConnectors={false}
>
  <path
    d=""
    style={
      Object {
        "fillOpacity": 0,
        "stroke": "#000",
        "strokeWidth": "2px",
      }
    }
  />
  <g
    key=".0"
    transform="translate(0 0)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
  <g
    key=".1"
    transform="translate(0 110)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
  <g
    key=".2"
    transform="translate(0 220)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
</VerticalLayout>
`;

exports[`VerticalLayout rendering with connectors 1`] = `
<VerticalLayout
  spacing={10}
  withConnectors={true}
>
  <path
    d="M10,60q0,-10 10,-10M130,60q0,-10 -10,-10M0,160c15,0 10,0 20,0M140,160c-15,0 -10,0 -20,0M10,260q0,10 10,10M130,260q0,10 -10,10M0,160q10,0 10,-10V60M140,160q-10,0 -10,-10V60M0,160q10,0 10,10V260M140,160q-10,0 -10,10V260"
    style={
      Object {
        "fillOpacity": 0,
        "stroke": "#000",
        "strokeWidth": "2px",
      }
    }
  />
  <g
    key=".0"
    transform="translate(20 0)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
  <g
    key=".1"
    transform="translate(20 110)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
  <g
    key=".2"
    transform="translate(20 220)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
</VerticalLayout>
`;
